res/controllers/Denon-MC7000-scripts.js

```diff
@@ -486,16 +510,31 @@ MC7000.PadButtons = function(channel, control, value, status, group) {
             midi.sendShortMsg(0x94 + deckNumber - 1, 0x14 + i, MC7000.padColor.rolloff);
         }
     } else if (MC7000.PADModeSavedLoop[deckNumber]) {
-        return;
+        if (value === 0x00) {
+            return; // don't respond to note off messages
+        }
+        i = control - 0x14;
+        engine.setValue(group, "beatloop_" + MC7000.fixedLoop[i] + "_toggle", true);
+        for (j =0; j < 8; j++) {
+            var activeLED = engine.getValue(group, "beatloop_" + MC7000.fixedLoop[j] + "_enabled") ? MC7000.padColor.fixedloopon : MC7000.padColor.fixedloopoff;
+            midi.sendShortMsg(0x94 + deckNumber - 1, 0x14 + j, activeLED);
+        }
     } else if (MC7000.PADModeSlicer[deckNumber]) {
         if (value > 0) {
-            var beats = 1 << (control % 4);
-            if (control > 0x17) {
-                engine.setValue(group, "beatjump_" + beats + "_backward", value);
+            i = control - 0x14; // unshifted button
+            j = control - 0x1C; // shifted button
+            if (control < 0x18) {
+                engine.setValue(group, "beatjump_" + MC7000.fixedLoop[i] + "_forward", true);
+                midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpFwd);
+            } else if (control > 0x17 && control < 0x1C) {
+                engine.setValue(group, "beatjump_" + MC7000.fixedLoop[i - 4] + "_backward", true);
                 midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpBack);
-            } else {
-                engine.setValue(group, "beatjump_" + beats + "_forward", value);
+            } else if (control > 0x1B && control < 0x20) {
+                engine.setValue(group, "beatjump_" + MC7000.fixedLoop[j + 4] + "_forward", true);
                 midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpFwd);
+            } else if (control > 0x1F) {
+                engine.setValue(group, "beatjump_" + MC7000.fixedLoop[j] + "_backward", true);
+                midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpBack);
```

Comment:
Can you explain how these numbers relate to the physical buttons of the controller?

Reply:
14 to 1B are the 8 Pad buttons
1C to 23 are the shifted pad buttons

The upper row 14 to 17 and shifted 1C to 1F are for jumping forward

The lower row 18 to 1B and shifted 20 to 23 are used to jump back

Reply:
ok, its currently hard to read from code. Can you change to more to something like this:
```javascript
// forward buttons 
if (control >= 0x14 && < 0x18) {
    engine.setValue(group, "beatjump_" + MC7000.fixedLoop[i] + "_forward", true);
    midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpFwd);
// backward buttons 
} else if (control >= 0x18 && control < 0x1C) {
    engine.setValue(group, "beatjump_" + MC7000.fixedLoop[i - 4] + "_backward", true);
    midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpBack);
// forward buttons (shifted controls)
} else if (control >= 0x1C && control < 0x20) {
    engine.setValue(group, "beatjump_" + MC7000.fixedLoop[j + 4] + "_forward", true);
    midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpFwd);
// backward buttons (shifted controls)
} else if (control >= 0x20 && control < 0x24) {
    engine.setValue(group, "beatjump_" + MC7000.fixedLoop[j] + "_backward", true);
    midi.sendShortMsg(0x94 + deckNumber - 1, control, MC7000.padColor.slicerJumpBack);
}
```

Reply:
I was not only talking about the comments. The expressions in the cases changed as well.

Reply:
Looked deeper into it and finally realized what you mean - obviously there is no difference but if you wanna use the real PAD button midi control numbers then it should read like 

```
(control >= 0x14 && <= 0x17)
(control >= 0x18 && control <= 0x1B)
(control >= 0x1C && control <= 0x1F)
(control >= 0x20 && control <= 0x23)
```
so that each 4 PADs in a row are readable from their midi controls inside the code.

Reply:
Thats fine too. doing the < 0x20 instead of <= 0x1F makes more sense if we may recieve floats for some reason, but that since we can be very sure they're ints, your proposal is fine too.